lib/rules/no-constant-binary-expression.js

```diff
@@ -0,0 +1,440 @@
+ * @returns {boolean} Will `node` always give the same result when compared to a
+ * static boolean value?
+ */
+function hasConstantStrictBooleanComparison(scope, node) {
+    switch (node.type) {
+        case "ObjectExpression": // Objects are not booleans
+        case "ArrayExpression": // Arrays are not booleans
+        case "ArrowFunctionExpression": // Functions are not booleans
+        case "FunctionExpression":
+        case "ClassExpression": // Classes are not booleans
+        case "NewExpression": // Objects are not booleans
+        case "TemplateLiteral": // Strings are not booleans
+        case "Literal": // True, false, or not boolean, literals never change.
+        case "UpdateExpression": // Numbers are not booleans
+            return true;
+        case "BinaryExpression":
+            return NUMERIC_OR_STRING_BINARY_OPERATORS.has(node.operator);
+        case "UnaryExpression": {
+            if (node.operator === "delete") {
+                return false;
+            }
+            if (node.operator === "!") {
+                return isConstant(scope, node.argument, true);
+            }
+
+            /*
+             * The remaining operators return either strings or numbers, neither
+             * of which are boolean.
+             */
+            return true;
+        }
+        case "SequenceExpression": {
+            const last = node.expressions[node.expressions.length - 1];
+
+            return hasConstantStrictBooleanComparison(scope, last);
+        }
+        case "Identifier":
+            return node.name === "undefined";
+        case "AssignmentExpression":
+            if (node.operator === "=") {
+                return hasConstantStrictBooleanComparison(scope, node.right);
+            }
+
+            /*
+             * Handling short-circuiting assignment operators would require
+             * walking the scope. We won't attempt that (for now...)
+             */
+            if (node.operator === "&&=" || node.operator === "||=" || node.operator === "??=") {
+                return false;
+            }
+
+            /*
+             * The remaining assignment expressions all result in either a number
+             * or a string, neither of which can ever be boolean.
+             */
+            return true;
+        case "JSXElement": // ESLint has a policy of not assuming any specific JSX behavior.
+        case "JSXFragment":
+            return false;
+        default:
+            return false;
+    }
+}
+
+/**
+ * Test if an AST node will always result in a newly constructed object
+ * @param {ASTNode} node The node to test
+ * @returns {boolean} Will `node` always be new?
+ */
+function isAlwaysNew(node) {
+    switch (node.type) {
+        case "ObjectExpression":
+        case "ArrayExpression":
+        case "ArrowFunctionExpression":
+        case "FunctionExpression":
+        case "ClassExpression":
+        case "NewExpression":
+            return true;
```

Comment:
Oh good catch, I didn't realize that! We could keep it but restrict it to non-shadowed built-ins. I would love to still be able to catch boxed primitive bugs like:

```
x = new Boolean(someVar) === true;
```

Just checking for the primitives: `String`, `Number`, `Boolean` might be enough, but we could also consider a large set of builtins like `Set`, `Array`, `Map`, `WeakMap` etc.

Reply:
I looked into this a bit more, and some of the non-primitive constructors (eg. `Promise`) don't work correctly with `isReferenceToGlobalVariable`. A non-shadowed construction `new Promise` does not have a resolved value:

```javascript
x === new Promise()
```

```javascript
const reference = scope.references.find(ref => ref.identifier === node);
// reference.resolved === null
```

I can look into the scope implementation to learn more in the next day or so, but I figured I'd flag here in case you had any ideas in the mean time.

Reply:
> I looked into this a bit more, and some of the non-primitive constructors (eg. `Promise`) don't work correctly with `isReferenceToGlobalVariable`. A non-shadowed construction `new Promise` does not have a resolved value:

Did you enable an es6+ environment? In test cases, it would look like this:

https://github.com/eslint/eslint/blob/8bb527660ffddc122204fb5bf59076687aaa78b8/tests/lib/rules/no-obj-calls.js#L32-L35

ESLint by default adds es5 globals only, so by default there is no `Promise` global and references to it would be unresolved.

Reply:
> We could keep it but restrict it to non-shadowed built-ins.

Makes sense to me. We could check whether or not the global is an ES built-in by using [`globals.builtin`](https://github.com/eslint/eslint/blob/8bb527660ffddc122204fb5bf59076687aaa78b8/lib/rules/no-extend-native.js#L56) like in the `no-extend-native` rule.